ykrt/src/compile/jitc_yk/opt/lvn.rs

```diff
@@ -0,0 +1,115 @@
+#[cfg(test)]
+use crate::compile::jitc_yk::jit_ir::ConsumeInst;
+use crate::compile::{
+    jitc_yk::jit_ir::{AddInst, Inst, InstIdx, Module, Operand},
+    CompilationError,
+};
+use typed_index_collections::TiVec;
+
+/// Perform (very very) simple dead-code elimination.
+pub(super) fn lvn(mut m: Module) -> Result<Module, CompilationError> {
+    // Which operations are used?
+    let mut used_ops = vec![false; m.len()];
+    for inst in m.iter_mut_insts() {
+        match inst {
+            Inst::LoadTraceInput(_) => (),
+            Inst::Add(x) => {
+                if let Operand::Local(y) = x.lhs() {
+                    used_ops[usize::from(y)] = true;
+                }
+                if let Operand::Local(y) = x.rhs() {
+                    used_ops[usize::from(y)] = true;
+                }
+            }
+            #[cfg(test)]
+            Inst::Consume(x) => {
+                if let Operand::Local(y) = x.operand() {
+                    used_ops[usize::from(y)] = true;
+                }
+            }
+            x => todo!("{x:?}"),
+        }
+    }
+
+    // Rewrite the sequence of instructions, eliminating instructions whose results are unused, and
+    // performing the necessary adjustments to operands in the light of those instructions being
+    // removed.
+    fn operand_off(op: Operand, offs: &Vec<InstIdx>) -> Operand {
+        match op {
+            Operand::Local(x) => Operand::Local(offs[usize::from(x)]),
+            Operand::Const(_) => op,
+        }
+    }
+
+    let mut off = 0;
+    let mut insts = TiVec::<InstIdx, _>::new();
+    let mut offs = Vec::new();
+    for (i, inst) in m.iter_insts().enumerate() {
+        // The `unwrap` can't fail because `i - off <= i` (by definition) and `i` is a
+        // representable `InstIdx`.
+        offs.push(InstIdx::new(i - off).unwrap());
```

Comment:
The code to update the offsets looks a bit tedious. I wonder how annoying this will be for more complex optimisations and whether there's a way to make this easier for the user (i.e. optimisation writer)?

Reply:
I also wonder that: it's point (5) above.